README.md
<!-- DO NOT EDIT README.md. It is created by README.Rmd -->

[![Build Status](https://travis-ci.org/jrnold/ggthemes.svg?branch=master)](https://travis-ci.org/jrnold/ggthemes)
[![rstudio mirror downloads](http://cranlogs.r-pkg.org/badges/ggthemes)](https://github.com/metacran/cranlogs.app)
[![cran version](http://www.r-pkg.org/badges/version/ggthemes)](http://cran.rstudio.com/web/packages/ggthemes)



# ggthemes


## Overview

Some extra geoms, scales, and themes for
[ggplot](http://ggplot2.org), including:

### Geoms

- ``geom_rangeframe`` : Tufte's range frame
- ``geom_tufteboxplot``: Tufte's box plot

### Themes 

- ``theme_base``: a theme resembling the default base graphics in R. See also ``theme_par``.
- ``theme_calc``: a theme based on LibreOffice Calc.
- ``theme_economist``: a theme based on the plots in the [The Economist](http://www.economist.com/) magazine.
- ``theme_excel``: a theme replicating the classic ugly gray charts in Excel
- ``theme_few``: theme from Stephen Few's  ["Practical Rules for Using Color in Charts"](http://www.perceptualedge.com/articles/visual_business_intelligence/rules_for_using_color.pdf).
- ``theme_fivethirtyeight``: a theme based on the plots at [fivethirtyeight.com](http://fivethirtyeight.com).
- ``theme_gdocs``: a theme based on Google Docs.
- ``theme_hc``: a theme based on [Highcharts JS](http://www.highcharts.com).
- ``theme_par``: a theme that uses the current values of the base graphics parameters in `par`.
- ``theme_pander``: a theme to use with the [pander](http://rapporter.github.io/pander/) package.
- ``theme_solarized``: a theme using the [solarized](http://ethanschoonover.com/solarized) color palette.
- ``theme_stata``: themes based on [Stata](http://www.stata.com/) graph schemes.
- ``theme_tufte``: a minimal ink theme based on Tufte's *The Visual Display of Quantitative Information*.
- ``theme_wsj``: a theme based on the plots in the [The Wall Street Journal](http://www.wsj.com/).

### Scales

- ``scale_colour_calc``, ``scale_shape_calc``: color and shape palettes from LibreOffice Calc.
- ``scale_colour_colorblind``: Colorblind safe palette from <http://jfly.iam.u-tokyo.ac.jp/color/>.
- ``scale_colour_economist``: colors used in plots in plots in *The Economist*.
- ``scale_colour_excel``: colors from new and old Excel.
- ``scale_colour_few``: color palettes from Stephen Few's ["Practical Rules for Using Color in Charts"](http://www.perceptualedge.com/articles/visual_business_intelligence/rules_for_using_color.pdf).
- ``scale_colour_gdocs``: color palette from Google Docs.
- ``scale_colour_hc``: a theme based on [Highcharts JS](http://www.highcharts.com).
- ``scale_colour_solarized``: [Solarized](http://ethanschoonover.com/solarized) colors
- ``scale_colour_stata``, ``scale_shapes_stata``, ``scale_linetype_stata``: color, shape, and linetype palettes from Stata graph schemes.
- ``scale_colour_tableau``, ``scale_shape_tableau``: color and shape palettes from [Tableau](http://www.tableau.com/).
- ``scale_colour_pander``, ``scale_fill_pander``: scales to use with the [pander](http://rapporter.github.io/pander/) package.
- ``scale_colour_ptol``, ``scale_fill_ptol``: color palettes from Paul Tol's [Colour Schemes](https://personal.sron.nl/~pault/colourschemes.pdf)
- ``scale_shape_cleveland``, ``scale_shape_tremmel``, ``scale_shape_circlefill``: shape scales from classic works in visual perception: Cleveland, Tremmel (1995), and Lewandowsky and Spence (1989).

Most of these scales also have associates palettes, as used  in the *scales* package.

### Miscellaneous

- `bank_slopes`: Find the optimal aspect ratio to bank slopes to 45 degrees

<!--  LocalWords:  geoms ggplot rangeframe Tufte's tufteboxplot calc
 -->
<!--  LocalWords:  LibreOffice Calc Few's fivethirtyeight gdocs hc JS
 -->
<!--  LocalWords:  Highcharts solarized stata Stata tufte wsj colour
 -->
<!--  LocalWords:  linetype cleveland tremmel circlefill Lewandowsky
 -->


## Install 

To install the stable version from CRAN,

```r
install.packages('ggthemes', dependencies = TRUE)
```

Or, to install the development version from github, use the
**devtools** package,

```r
library("devtools")
install_github(c("hadley/ggplot2", "jrnold/ggthemes"))
```


<!--  LocalWords:  CRAN 'ggthemes' github devtools jrnold ggthemes
 -->
<!--  LocalWords:  Rtools
 -->


## Contribute

Contributions are welcome! If you would like to add a theme, scales,
etc., fork the repository, add your theme, and submit a pull request.


# Examples


```r
library("ggplot2")
library("ggthemes")

p <- ggplot(mtcars, aes(x = wt, y = mpg)) +
  geom_point() +
  ggtitle("Cars")

p2 <- ggplot(mtcars, aes(x = wt, y = mpg, colour = factor(gear))) +
  geom_point() +
  ggtitle("Cars")

p3 <- p2 + facet_wrap(~ am)
```

### Tufte theme and geoms

Minimal theme and geoms based on plots in *The Visual Display of Quantitative Information*.


```r
p + geom_rangeframe() +
  theme_tufte() + 
  scale_x_continuous(breaks = extended_range_breaks()(mtcars$wt)) +
  scale_y_continuous(breaks = extended_range_breaks()(mtcars$mpg))
```

![plot of chunk tufte-rangeframe](figure/tufte-rangeframe-1.png)

The function `geom_tufteboxplot` creates several variants of Tufte's minimal-ink boxplots.
For a boxplot with a point indicating the median, a gap indicating the interquartile range,
and lines for whiskers:

```r
p4 <- ggplot(mtcars, aes(factor(cyl), mpg))

p4 + theme_tufte(ticks=FALSE) + geom_tufteboxplot()
```

![plot of chunk tufteboxplot](figure/tufteboxplot-1.png)

For a boxplot with an offset line indicating the interquartile range and a gap indicating the median:

```r
p4 + theme_tufte(ticks=FALSE) +
  geom_tufteboxplot(median.type = "line")
```

![plot of chunk tufteboxplot2](figure/tufteboxplot2-1.png)

For a boxplot with an line indicating the interquartile range, a gap indicating the median, and 
points indicating the minimum and maximum:

```r
p4 + theme_tufte(ticks=FALSE) +
  geom_tufteboxplot(median.type = "line", whisker.type = 'point', hoffset = 0)
```

![plot of chunk tufteboxplot3](figure/tufteboxplot3-1.png)

For a boxplot with a wide line indicating the interquartile range, a gap indicating the median, and 
lines indicating the minimum and maximum

```r
p4 + theme_tufte(ticks=FALSE) +
  geom_tufteboxplot(median.type = "line", whisker.type = 'line', hoffset = 0, width = 3)
```

```
## Warning: position_dodge requires non-overlapping x intervals
```

![plot of chunk tufteboxplot4](figure/tufteboxplot4-1.png)

### Economist theme

A theme that approximates the style of plots in The Economist
magazine.


```r
p2 + theme_economist() + scale_colour_economist() 
```

![plot of chunk economist](figure/economist-1.png)

### Solarized theme

A theme and color and fill scales based on the Solarized palette.

The light theme.


```r
p2 + theme_solarized() +
  scale_colour_solarized("blue")
```

![plot of chunk solarized-light](figure/solarized-light-1.png)

The dark theme.


```r
p2 + theme_solarized(light = FALSE) +
  scale_colour_solarized("red")
```

![plot of chunk solarized-dark](figure/solarized-dark-1.png)

An alternative theme.


```r
p2 + theme_solarized_2(light = FALSE) +
  scale_colour_solarized("blue")
```

![plot of chunk solarized-alt](figure/solarized-alt-1.png)


### Stata theme 

Themes and scales (color, fill, linetype, shapes) based on the graph
schemes in Stata.


```r
p2 + theme_stata() + scale_colour_stata()
```

![plot of chunk stata](figure/stata-1.png)

### Excel 2003 theme

For that classic ugly look and feel. For ironic purposes only. 3D bars
and pies not included. Please never use this theme.


```r
p2 + theme_excel() + scale_colour_excel()
```

![plot of chunk excel1](figure/excel1-1.png)


```r
ggplot(diamonds, aes(x = clarity, fill = cut)) +
  geom_bar() +
  scale_fill_excel() +
  theme_excel()
```

![plot of chunk excel2](figure/excel2-1.png)

### Inverse Gray Theme

Inverse of `theme_gray`, i.e. white plot area and gray background.


```r
p2 + theme_igray()
```

![plot of chunk igray](figure/igray-1.png)

### Fivethirtyeight theme

Theme and color palette based on the plots at [fivethirtyeight.com](http://fivethirtyeight.com).


```r
p2 + geom_smooth(method = "lm", se = FALSE) +
  scale_color_fivethirtyeight("cyl") +
  theme_fivethirtyeight()
```

![plot of chunk fivethirtyeight](figure/fivethirtyeight-1.png)

### Paul Tol Scales

Color palette based on [Paul Tol's Colour Schemes](https://personal.sron.nl/~pault/colourschemes.pdf).


```r
p2 + geom_smooth(method = "lm", se = FALSE) +
  scale_color_ptol("cyl") +
  theme_minimal()
```

![plot of chunk paultol](figure/paultol-1.png)


```r
ggplot(diamonds, aes(x = clarity, fill = cut)) +
  geom_bar() +
  scale_fill_ptol() +
  theme_minimal()
```

![plot of chunk paultol-barplot](figure/paultol-barplot-1.png)

### Tableau Scales

Color, fill, and shape scales based on those used in the Tableau software.


```r
p2 + theme_igray() + scale_colour_tableau()
```

![plot of chunk tableau](figure/tableau-1.png)


```r
p2 + theme_igray() + scale_colour_tableau("colorblind10")
```

![plot of chunk tableau-colorbind10](figure/tableau-colorbind10-1.png)

### Stephen Few's Practical Rules for Using Color ...

Color palette and theme based on Stephen Few's ["Practical Rules for Using Color in Charts"](http://www.perceptualedge.com/articles/visual_business_intelligence/rules_for_using_color.pdf).


```r
p2 + theme_few() + scale_colour_few()
```

![plot of chunk few](figure/few-1.png)

### Wall Street Journal

Theme and some color palettes based on plots in the *The Wall Street Journal*.


```r
p2 + theme_wsj() + scale_colour_wsj("colors6", "")
```

![plot of chunk wsj](figure/wsj-1.png)

### Base and Par Themes

Theme that resembles the default theme in the `base` graphics in R.


```r
p2 + theme_base()
```

![plot of chunk unnamed-chunk-5](figure/unnamed-chunk-5-1.png)

### Par theme

Theme that uses the current values of base graphics stored in  `par`. Not all `par` parameters, are supported, and not all are relevant to **ggplot2** themes.


```r
par(fg = "blue", bg = "gray", col.lab = "red", font.lab = 3)
p2 + theme_par()
```

![plot of chunk unnamed-chunk-6](figure/unnamed-chunk-6-1.png)


### GDocs Theme

Theme and color palettes based on the defaults in Google Docs.


```r
p2 + theme_gdocs() + scale_color_gdocs()
```

![plot of chunk gdocs](figure/gdocs-1.png)

### Calc Theme

Theme and color and shape palettes based on the defaults in LibreOffice Calc.


```r
p2 + theme_calc() + scale_color_calc()
```

![plot of chunk calc](figure/calc-1.png)

### Pander Theme

Theme and color palettes based on the [pander package](http://rapporter.github.io/pander/).


```r
p2 + theme_pander() + scale_colour_pander()
```

![plot of chunk pander-scatterplot](figure/pander-scatterplot-1.png)


```r
ggplot(diamonds, aes(x = clarity, fill = cut)) +
  geom_bar() +
  theme_pander() +
  scale_fill_pander()
```

![plot of chunk pander-barplot](figure/pander-barplot-1.png)

### Highcharts  theme

A theme that approximates the style of plots in [Highcharts JS](http://www.highcharts.com/demo).


```r
p2 + theme_hc() + scale_colour_hc()
```

![plot of chunk hc-default](figure/hc-default-1.png)

```r
p2 + theme_hc(bgcolor = "darkunica") +
  scale_colour_hc("darkunica")
```

![plot of chunk hc-darkunica](figure/hc-darkunica-1.png)


```r
dtemp <- data.frame(months = factor(rep(substr(month.name,1,3), 4), levels = substr(month.name,1,3)),
                    city = rep(c("Tokyo", "New York", "Berlin", "London"), each = 12),
                    temp = c(7.0, 6.9, 9.5, 14.5, 18.2, 21.5, 25.2, 26.5, 23.3, 18.3, 13.9, 9.6,
                             -0.2, 0.8, 5.7, 11.3, 17.0, 22.0, 24.8, 24.1, 20.1, 14.1, 8.6, 2.5,
                             -0.9, 0.6, 3.5, 8.4, 13.5, 17.0, 18.6, 17.9, 14.3, 9.0, 3.9, 1.0,
                             3.9, 4.2, 5.7, 8.5, 11.9, 15.2, 17.0, 16.6, 14.2, 10.3, 6.6, 4.8))
```


```r
ggplot(dtemp, aes(x = months, y = temp, group = city, color = city)) +
  geom_line() +
  geom_point(size = 1.1) + 
  ggtitle("Monthly Average Temperature") +
  theme_hc() +
  scale_colour_hc()
```

![plot of chunk hc-default-line](figure/hc-default-line-1.png)


```r
ggplot(dtemp, aes(x = months, y = temp, group = city, color = city)) +
  geom_line() + 
  geom_point(size = 1.1) + 
  ggtitle("Monthly Average Temperature") +
  theme_hc(bgcolor = "darkunica") +
  scale_fill_hc("darkunica")
```

![plot of chunk hc-darkunica-line](figure/hc-darkunica-line-1.png)

## Maps theme

A theme useful for displaying maps.


```r
library("maps")
us <- fortify(map_data("state"), region = "region")
ggplot() +
  geom_map(data  =  us, map = us,
           aes(x = long, y = lat, map_id = region, group = group),
           fill = "white", color = "black", size = 0.25) +
  coord_map("albers", lat0 = 39, lat1 = 45) +
  theme_map()
```

```
## Warning: Ignoring unknown aesthetics: x, y
```

![plot of chunk map](figure/map-1.png)



<!--  LocalWords:  dsamp ggplot2 ggthemes nrow Tufte geoms tufte aes
 -->
<!--  LocalWords:  rangeframe ggplot mtcars boxplot tufteboxplot cyl
 -->
<!--  LocalWords:  qplot colour ggtitle Solarized solarized Stata 3D
 -->
<!--  LocalWords:  linetype stata excel1 excel2 igray Fivethirtyeight
 -->
<!--  LocalWords:  fivethirtyeight lm se colorbind10 colorblind10 wsj
 -->
<!--  LocalWords:  Few's colors6 GDocs gdocs Calc LibreOffice calc JS
 -->
<!--  LocalWords:  scatterplot barplot Highcharts hc darkunica dtemp
 -->
<!--  LocalWords:  bgcolor substr
 -->


NEWS
ggthemes 3.3.0
------------------------

* Update themes to changes in ggplot 2.1.0.9000. Thanks @juliasilge (#71)
* `tufte_boxplot` uses `position="dodge"` by default. Thanks @jgellar (#68)
* Bugfix: Fix errors in `geom_rangeframe` in new version of ggplot2. Thanks @coulmont (#70)

ggthemes 3.2.0
------------------------

* Bugfix: In function `bank_slopes`, remove methods `ao`, `gor`, and `lor`. These methods were not producing reliable results, and should not produce results much different thatn `ms` or `as`. If used, they will produce a warning, and `ms` will be used instead. (#68)

ggthemes 3.1.0
------------------------

* Add `ptol_pal`, `scale_colour_ptol`, and `scale_file_ptol` based on Paul Tol's qualitative color palettes. Thanks @jmlondon. (#65)

ggthemes 3.0.4
------------------------

* Fix panel.grid.major.y colors for `theme_hc`. Thanks @flying-sheep! (#64)

ggthemes 3.0.3
------------------------

* Bugfix: remove `lintr` tests that broke with new `testthat`.

ggthemes 3.0.2
------------------------

* Bugfix: export `ggthemes_data`

ggthemes 3.0.1
------------------------

* Bugfix: Remove border around legends in `theme_gdocs`. Thanks SandyMuspratt! Pull request #53.
* Bugfix: Remove border around legends in `theme_calc`.
* Bugfix: Add colour and fill values to `theme_foundation`. This restores its behavior to version 2.2.1. Fixes Issue #56.

ggthemes 3.0.0
------------------------

* For `geom_tufteboxplot` and `GeomTufteboxplot`. Option `median.type` supports only
  `line` and `point` options. Added option `whisker.type` which allows for whiskers
  to be specified by lines or points.
* Changed default stat for `geom_tufteboxplot` to `stat_fivenumber`.
* Added `theme_base` and `theme_par`
* Removed `scale_(x|y)_tufte`. See issue #49
* `theme_foundation` rewritten. Removed `use_sizes` argument.
* Import rather than depend on `ggplot2`
* Update geoms, scales, and themes to ggplot >= 2.0.0


ggthemes 2.2.1
------------------------

* bugfix: fix R CMD check notes due to change in how it handles non-base default packages: http://developer.r-project.org/blosxom.cgi/R-devel/2015/06/29#n2015-06-29
* bugfix: fix bad URLs

ggthemes 2.2.0
------------------------

* Added `theme_map` from hrbrmstr

ggthemes 2.1.1
------------------------

* bugfix: use title case in package title; refer to ggplot2 in title
* bugfix: fix non-escaped URLs in colorblind.R
* bugfix: fix broken URL in few.R
* bugfix: remove non-ascii characters in shapes.R
* bugfix: use `requireNamespace()` instead of `require()`

ggthemes 2.1.0
------------------------

* New theme and scale based on Highcharts JS: `scale_*_hc`, `theme_hc`. Thanks jbkunst!

ggthemes 2.0.0
------------------------

* New options for `geom_tufteboxplot`. `median.type` takes values "point", "box", or "line". Option `usebox` is no longer supported, and equivalent to `median.type="box"`. Thanks weitzner! Issue #13, PR #19.
* All theme functions have `base_size` and `base_family` as first two arguments. This affects `theme_economist_white`, `theme_excel`, `theme_pander`, `theme_solid`, and `theme_tufte`. Thanks ptoche! Issue #25.
* `theme_pander`: Options `ff` and `fs` are deprecated; use `base_family` and `base_size` options instead, respectively.

ggthemes 1.9.0
------------------------

* added vignette
* added themes / scales based on factory defaults of pander package: `pander_palette`, `scale_color_pander`, `scale_fill_pander`, `theme_pander`. Thanks daroczig.
* added additional fill and gradient color scales: `scale_fill_fivethirtyeight`, `scale_colour_gradient_tableau`, `scale_fill_gradient_tableau`, `scale_fill_continuous_tableau`, `scale_colour_gradient2_tableau`, `scale_fill_gradient2_tableau`. Thanks bowerth.

ggthemes 1.8.0
------------------------

* added themes and color scale from fivethirtyeight.com: `theme_fivethirtyeight`, `fivethirtyeight_pal`, `scale_color_fivethirtyeight`

ggthemes 1.7.0
------------------------

* added a new breaks algorithm: `scales_extended_range_breaks`, `extended_range_breaks`
* added scales which implement the breaks algorithm: `scale_x_tufte`, `scale_y_tufte`
* added new label format: `smart_digits`, `smart_digits_format`

ggthemes 1.6.1
------------------------

* fix bug in `theme_wsj` (issue #17)

ggthemes 1.6.0
------------------------

* added `theme_solid`

ggthemes 1.5.1
------------------------

* fix failures in R CMD check

ggthemes 1.5.0
------------------------

* added alternative Solarized theme: `theme_solarized_2`
* `theme_solarized`: adjusted the colors
* added LibreOffic Calc themes and palettes: `theme_calc`, `calc_pal`, `scale_fill_calc`, `scale_colour_calc`, `cacl_shape_pal`, `scale_shape_calc`
* added Google Docs themes and palettes: `theme_gdocs`, `gdocs_pal`, `scale_fill_gdocs`, `scale_colour_gdocs`
* fixed some examples

ggthemes 1.4.0
------------------------

* fix bug in default palettes of `scale_colour_excel` and `scale_fill_excel`.

ggthemes 1.3.4
------------------------

* added Tableau sequential colour palettes: `tableau_seq_gradient_pal`, `scale_colour_gradient_tableau`, `scale_fill_gradient_tableau`.
* added Tableau diverging colour palettes: `tableau_div_gradient_pal`, `scale_colour_gradient2_tableau`, `scale_fill_gradient2_tableau`.
* `tableau_colour_pal`: added palette "cyclical


ggthemes 1.3.3
------------------------

* fix issue #11: error with fonts and `theme_wsj` in Windows

ggthemes 1.3.1
------------------------

* bank_slopes returns y/x aspect ratio to be compatible with coord_fixed
* improved documentation

ggthemes 1.3.0
------------------------

* renamed `theme_excel2003` to `theme_excel`
* replace `scale_fill_excel2003` and `scale_fill_excel10` with `scale_fill_excel`
* replaced `scale_colour_excel2003` and `scale_colour_excel10` with `scale_colour_excel`
* replaced `excel2003_pal` and `excel10_pal` with `excel_pal`
* renamed `theme_base` to `theme_foundation`

ggthemes 1.2.0
------------------------


R/base.R
#' Theme Base
#'
#' Theme similar to the default settings of the \sQuote{base} R graphics.
#'
#' @inheritParams ggplot2::theme_bw
#' @export
#' @family themes
#' @examples
#' library("ggplot2")
#' p <- ggplot(mtcars) + geom_point(aes(x = wt, y = mpg,
#'      colour=factor(gear))) + facet_wrap(~am)
#' p + theme_base()
theme_base <- function(base_size = 16, base_family = "") {
 theme_foundation() +
    theme(line = element_line(colour = "black",
                              lineend = "round",
                              linetype = "solid"),
          rect = element_rect(fill = "white",
                              colour = "black",
                              linetype = "solid"),
          text = element_text(colour = "black",
                              face = "plain",
                              family = base_family,
                              size = base_size,
                              vjust = 0.5,
                              hjust = 0.5,
                              lineheight = 1),
          panel.grid = element_blank(),
          strip.background = element_rect(colour = NA),
          legend.key = element_rect(colour = NA),
          title = element_text(size = rel(1)),
          plot.title = element_text(size = rel(1.2), face = "bold"),
          strip.text = element_text(),
          axis.ticks.length = unit(0.5, "lines")
          )
  # TODO: get margins right
}

# Notes for generating a theme that uses par() for its values.
#
# $xlog
# [1] TRUE
#
# $ylog
# [1] TRUE
#
# Justification of strings in text, mtext, and title
# # text = element_text(vjust = par()$adj, hjust = par$adj())
# $adj
# [1] 0.5
#
# $ann
# [1] TRUE
#
# $ask
# [1] FALSE
#
# Background. rect = element_rect(fill = par()$bg)
# $bg
# [1] "white"
#
# # Type of box drawn around the plot
# # Which sides of the box to draw
# $bty
# [1] "o"
#
# # magnification of text and symbols relative to the default. ggplot uses base_size instead.
# $cex
# [1] 1
#
# # mag of axis relative to current setting of cex
# # axis.text = element_text(size = rel(par()$cex.axis))
# $cex.axis
# [1] 1

# # mag of axis relative to current setting of cex
# # axis.title = element_text(size = rel(par()$cex.lab))
# $cex.lab
# [1] 1
#
# # magnification of plot title relative to cex
# # plot.title = element_text(size = rel(par()$cex.main))
# $cex.main
# [1] 1.2
#
# # ggplot does not have subtitles
# # magnification of subtitle relative to cex
# # strip.title = element_text(size = rel(par()$cex.sub))
# $cex.sub
# [1] 1
#
# $cin
# # character size in inches
# [1] 0.2000000 0.2666667
#
# # default plotting color - not part of theme in gggplot
# $col
# [1] "black"
#
# # color for axis annotation
# #
# $col.axis
# [1] "black"
# # color for x and y labels
# # axis.text = element_text(colour = par()$col.axis)
#
# # color for x and y labels
# # axis.title = element_text(colour = par()$col.lab)
# $col.lab
# [1] "black"
#
# # color for main titles
# # plot.title = element_text(colour = par()$col.main)
# $col.main
# [1] "black"
#
# # color for subtitles
# # strip.title = element_text(colour = par()$col.sub)
# $col.sub
# [1] "black"
#
# # size of default character
# $cra
# [1] 14.4 19.2
#
# # numerical values for how single characters rotated. Nothing similar in ggplot
# $crt
# [1] 0
#
# # size of default characters in inches
# # Is this base size?
# $csi
# [1] 0.2666667
#
# # size of default character in user coord
# # not relevant
# $cxy
# [1] 0.1859782 0.3665854
#
# # device dimensions. not relevant
# $din
# [1] 11.236111  8.847222
#
# # error reporting. not relevant
# $err
# [1] 0
#
# # default font family
# # base_family = par()$family
# $family
# [1] ""
#
# # color of foreground in plots. Used in axes and boxes around plots.
# line = element_line(colour = par()$fg)
# rect = element_rect(colour = par()$fg)
# text = = element_text(colour = par()$fg)
# $fg
# [1] "black"
#
# # gives NDC coordinates of figure region in display device
# $fig
# [1] 0 1 0 1
#
# # figure region dimensions in inches
# # TODO: use for aspect ratio?
# $fin
# [1] 11.236111  8.847222
#
# # which font to use for text.
# # 1 = "plain"
# # 2 = "bold"
# # 3 = "italic"
# # 4 = "bold.italic"
# text = element_text(face = c("plain", "bold", "italic", "bold.italic")[par()$font])
# $font
# [1] 1
#
# # font to use for axis
# axis.text = element_text(face = c("plain", "bold", "italic", "bold.italic")[par()$font])
# $font.axis
# [1] 1
#
# axis.title = element_text(face = c("plain", "bold", "italic", "bold.italic")[par()$font])
# $font.lab
# [1] 1
#
# axis.title = element_text(face = c("plain", "bold", "italic", "bold.italic")[par()$font])
# $font.main
# [1] 2
#
# strip.title = element_text(face = c("plain", "bold", "italic", "bold.italic")[par()$font])
# $font.sub
# [1] 1
#
# # default number of tick-marks in x and y, and label lenghth.
# Not sure how that can be used
# $lab
# [1] 5 5 7
#
# # style of axis labels.
# # TODO: code that sets axis.text.x and axis.text.y angle according to its values.
# $las
# [1] 0
#
# # line end style
# line = element_line(lineend = par()$lend)
# $lend
# [1] "round"
#
# # line height
# text = element_text(lineheight = par()$lheight * par()$??)
# $lheight
# [1] 1
#
# # Line join style
# # not sure how this is used in ggplot
# $ljoin
# [1] "round"
#
# # line mitre imit. Not used in ggplot2.
# $lmitre
# [1] 10
#
# # Line type
# # line = element_line(linetype = par()$lty)
# $lty
# [1] "solid"
#
# # Line width?
# # Does this set size? ??
# # Maybe: line = element_line(size = par()$lwd)
# $lwd
# [1] 1
#
# # margin size in inches
# plot.margin = par()$mai
# $mai
# [1] 1.360000 1.093333 1.093333 0.560000
#
# # Number of lines of margin. How is this different than mai?
# $mar
# [1] 5.1 4.1 4.1 2.1
#
# $mex
# [1] 1
#
# # changes layout. ggplot uses facets.
# $mfcol
# [1] 1 1
#
# # used for layout
# $mfg
# [1] 1 1 1 1
#
# # changes layout. ggplot uses facets.
# $mfrow
# [1] 1 1
#
# # margine line in mex units for axis title, axis labels, and axis.line
# $mgp
# [1] 3 1 0
#
# # ignored in R
# $mkh
# [1] 0.001
#
# # irrelevant to ggplot
# $new
# [1] FALSE
#
# # size of outer margins in lines of text
# # TODO: what is this in ggplot
# $oma
# [1] 0 0 0 0
#
# # regions inside out margins in NDC
# # TODO: ?
# $omd
# [1] 0 1 0 1
#
# # size of outer margins in inches
# # TODO???
# $omi
# [1] 0 0 0 0
#
# # irrelevant
# $page
# [1] TRUE
#
# # dfault plotting symbol. not in themes.
# $pch
# [1] 1
#
# # Plot dimensions in inches.
# $pin
# [1] 9.582778 6.393889
#
# # Plot region as fractions of current figure region.
# $plt
# [1] 0.09730532 0.95016069 0.15372057 0.87642072
#
# # point size of text (not symbols)
# base_size = par()$ps
# $ps
# [1] 16
#
# # type of region to be drawn. s = square. m = maximal.
# # not sure how this maps to ggplot
# $pty
# [1] "m"
#
# # not used
# $smo
# [1] 1
#
# # string rotation in degrees
# # this is for text() plots not titles and labels
# $srt
# [1] 0
#
# # length of tick marks
# # use this to set axis.ticks.length
# $tck
# [1] NA
#
# # length of tick marks
# $tcl
# [1] -0.5
#
# # extremes of user coord of plotting regsion
# $usr
# [1] -0.0381697  0.9924122  0.2730712  1.0279588
#
# # used for generating ticks
# $xaxp
# [1]  1 10  3
#
# # used to generate axis
# $xaxs
# [1] "r"
#
# # any values other than "n" implies plotting x axis.
# $xaxt
# [1] "s"
#
# $xpd
# [1] FALSE
#
# $yaxp
# [1]  2 10 -4
#
# $yaxs
# [1] "r"
#
# # any value other than "n" implies plotting y axis.
# $yaxt
# [1] "s"
#
# # positioning of text in margins by axis and mtext.
# $ylbias
# [1] 0.2

# #' Theme Par
#'
#' Theme which takes its values from the current \sQuote{base} graphics
#' parameter values in \code{\link{par}}.
#'
#' Currently this theme uses the values of the parameters:
#' \code{"code"}, "\code{"ps"}", \code{"code"} \code{"family"}, \code{"fg"},
#' \code{"bg"}, \code{"adj"}, \code{"font"}, \code{"cex.axis"},
#' \code{"cex.lab"}, \code{"cex.main"}, \code{"cex.sub"}, \code{"col.axis"},
#' \code{"col.lab"}, \code{"col.main"}, \code{"col.sub"}, \code{"font"},
#' \code{"font.axis"}, \code{"font.lab"}, \code{"font.main"},
#' \code{"font.sub"}, \code{"las"}, \code{"lend"},
#' \code{"lheight"}, \code{"lty"}, \code{"mar"}, \code{"ps"}, \code{"tcl"},
#' \code{"tck"}, \code{"xaxt"}, \code{"yaxt"}.
#'
#' This theme does not translate the base graphics perfectly, so the graphs
#' produced by it will not be identical to those produced by base graphics,
#' most notably in the spacing of the margins.
#'
#' @inheritParams ggplot2::theme_bw
#' @export
#' @family themes
#' @examples
#' library("ggplot2")
#' p <- ggplot(mtcars) + geom_point(aes(x = wt, y = mpg,
#'      colour=factor(gear))) + facet_wrap(~am)
#' par(font = 2, col.lab = "red", fg = "blue")
#' p + theme_par()
theme_par <- function(base_size = par()$ps, base_family = par()$family) {
  faces <- c("plain", "bold", "italic", "bold.italic")
  half_line <- base_size / 2
  thm <- theme_foundation() %+replace%
        theme(line = element_line(colour = par()$fg,
                                   size = 0.5,
                                   lineend = par()$lend,
                                   linetype = par()$lty),
        rect = element_rect(fill = par()$bg,
                            colour = par()$fg,
                            size = 0.5,
                            linetype = par()$lty),
        text = element_text(colour = par()$fg,
                            face = faces[par()$font],
                            family = base_family,
                            size = base_size,
                            angle = 0,
                            margin = margin(),
                            vjust = par()$adj,
                            hjust = par()$adj,
                            lineheight = par()$lheight,
                            debug = FALSE),
        axis.title = element_text(size = rel(par()$cex.lab),
                                  colour = par()$col.lab,
                                  face = faces[par()$font.lab]),
        axis.text = element_text(size = rel(par()$cex.axis),
                                 colour = par()$col.axis,
                                 face = faces[par()$font.axis]),
        axis.text.x = element_text(margin = margin(t = 0.8 * half_line / 2,
                                                   b = 0.8 * half_line / 2)),
        axis.text.y = element_text(margin = margin(r = 0.8 * half_line / 2,
                                                   l = 0.8 * half_line / 2)),
        axis.ticks = element_line(colour = par()$fg),
        legend.title = element_text(colour = par()$fg),
        legend.text = element_text(colour = par()$fg),
        legend.spacing = unit(0.2, "cm"),
        legend.key = element_rect(colour = NA),
        panel.spacing = unit(half_line, "pt"),
        panel.spacing.x = NULL,
        panel.spacing.y = NULL,
        panel.background = element_rect(fill = NA, colour = par()$col),
        panel.grid = element_blank(),
        plot.background = element_rect(colour = NA),
        plot.margin = unit(par()$mar, "lines"),
        plot.title = element_text(size = rel(par()$cex.main),
                                  face = faces[par()$font.main],
                                  colour = par()$col.main,
                                  margin = margin(b = half_line * 1.2)),
        strip.text = element_text(size = rel(par()$cex.sub),
                                  face = faces[par()$font.sub],
                                  colour = par()$col.sub),
        strip.text.x = element_text(margin = margin(t = half_line,
                                                    b = half_line)),
        strip.text.y = element_text(margin = margin(l = half_line,
                                                    r = half_line)),
        strip.background = element_rect(colour = NA)
  )

  las <- par()$las
  if (las == 0) {
    # parallel to axis
    thm <- thm + theme(axis.title.x = element_text(angle = 0),
                       axis.title.y = element_text(angle = 90))
  } else if (las == 1) {
    # horizontal
    thm <- thm + theme(axis.title.x = element_text(angle = 0),
                       axis.title.y = element_text(angle = 0))
  } else if (las == 2) {
    # perpendicular
    thm <- thm + theme(axis.title.x = element_text(angle = 90),
                       axis.title.y = element_text(angle = 0))
  } else if (las == 3) {
    # vertical
    thm <- thm + theme(axis.title.x = element_text(angle = 90),
                       axis.title.y = element_text(angle = 90))
  }

  # ticks
  if (! is.na(par()$tck)) {
    thm <- thm + theme(axis.ticks.length = unit(- par()$tck, "snpc"))
  } else {
    thm <- thm + theme(axis.ticks.length = unit(- par()$tcl, "lines"))
  }
  # plot x or y axis
  if (par()$xaxt == "n") {
    thm <- thm + theme(axis.line.x = element_blank(),
                       axis.text.x = element_blank(),
                       axis.ticks.x = element_blank())
  }
  if (par()$yaxt == "n") {
    thm <- thm + theme(axis.line.y = element_blank(),
                       axis.text.y = element_blank(),
                       axis.ticks.y = element_blank())
  }

  thm
  # TODO: get margins right
}
